File: backend/app/shared/domain/reports.py
"""Pure, framework-free analytics over transactions. No I/O, no framework
imports (mirrors `shared/domain/budgeting`). The TS core in
`frontend/src/shared/domain/reports.ts` must agree on the same numbers — the
mirrored tests assert identical values on a fixed dataset.

Everything is computed from a normalized transaction list so the same code runs
server-side (over SQL rows) and client-side (over Dexie rows). Transfers are
internal moves and are excluded from income/expense analytics.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class ReportTx:
    type: str  # income | expense | transfer
    amount: float
    date: str  # YYYY-MM-DD
    category_id: str | None = None
    merchant_id: str | None = None


@dataclass
class MonthPoint:
    month: str  # YYYY-MM
    income: float = 0.0
    expense: float = 0.0
    net: float = 0.0


@dataclass
class KeyTotal:
    key: str
    total: float


@dataclass
class Report:
    by_month: list[MonthPoint] = field(default_factory=list)
    spending_by_category: list[KeyTotal] = field(default_factory=list)
    spending_by_merchant: list[KeyTotal] = field(default_factory=list)
    largest_expenses: list[ReportTx] = field(default_factory=list)
    income_total: float = 0.0
    expense_total: float = 0.0
    net: float = 0.0
    savings_rate: float = 0.0
    health_score: int = 0


def savings_rate(income_total: float, expense_total: float) -> float:
    """Fraction of income kept. 0 when there is no income."""
    if income_total <= 0:
        return 0.0
    return (income_total - expense_total) / income_total


def health_score(rate: float) -> int:
    """Financial-health score in [0, 100], from the savings rate.

    A savings rate of 30%+ is full marks; 0% or negative is 0. Linear between.
    """
    scaled = round((rate / 0.30) * 100)
    return max(0, min(100, scaled))


def _sorted_totals(totals: dict[str, float]) -> list[KeyTotal]:
    return [
        KeyTotal(key=k, total=v)
        for k, v in sorted(totals.items(), key=lambda kv: (-kv[1], kv[0]))
    ]


def build_report(txs: list[ReportTx], largest_n: int = 5) -> Report:
    months: dict[str, MonthPoint] = {}
    by_category: dict[str, float] = {}
    by_merchant: dict[str, float] = {}
    income_total = 0.0
    expense_total = 0.0
    expenses: list[ReportTx] = []

    for tx in txs:
        if tx.type == "transfer":
            continue  # internal move, not spending or income
        month = tx.date[:7]
        point = months.setdefault(month, MonthPoint(month=month))
        if tx.type == "income":
            income_total += tx.amount
            point.income += tx.amount
        elif tx.type == "expense":
            expense_total += tx.amount
            point.expense += tx.amount
            expenses.append(tx)
            if tx.category_id is not None:
                by_category[tx.category_id] = by_category.get(tx.category_id, 0.0) + tx.amount
            if tx.merchant_id is not None:
                by_merchant[tx.merchant_id] = by_merchant.get(tx.merchant_id, 0.0) + tx.amount
        point.net = point.income - point.expense

    rate = savings_rate(income_total, expense_total)
    return Report(
        by_month=[months[m] for m in sorted(months)],
        spending_by_category=_sorted_totals(by_category),
        spending_by_merchant=_sorted_totals(by_merchant),
        largest_expenses=sorted(expenses, key=lambda t: (-t.amount, t.date))[:largest_n],
        income_total=income_total,
        expense_total=expense_total,
        net=income_total - expense_total,
        savings_rate=rate,
        health_score=health_score(rate),
    )

File: frontend/src/shared/domain/reports.ts
/**
 * Pure, framework-free analytics over transactions — the TS mirror of
 * `backend/app/shared/domain/reports.py`. Both cores must agree on the same
 * numbers (the mirrored tests assert identical values on a fixed dataset).
 *
 * Transfers are internal moves and are excluded from income/expense analytics.
 */

export interface ReportTx {
  type: "income" | "expense" | "transfer";
  amount: number;
  date: string; // YYYY-MM-DD
  categoryId?: string | null;
  merchantId?: string | null;
}

export interface MonthPoint {
  month: string; // YYYY-MM
  income: number;
  expense: number;
  net: number;
}

export interface KeyTotal {
  key: string;
  total: number;
}

export interface Report {
  byMonth: MonthPoint[];
  spendingByCategory: KeyTotal[];
  spendingByMerchant: KeyTotal[];
  largestExpenses: ReportTx[];
  incomeTotal: number;
  expenseTotal: number;
  net: number;
  savingsRate: number;
  healthScore: number;
}

export function savingsRate(incomeTotal: number, expenseTotal: number): number {
  if (incomeTotal <= 0) return 0;
  return (incomeTotal - expenseTotal) / incomeTotal;
}

/** Financial-health score in [0, 100] from the savings rate (30%+ = full). */
export function healthScore(rate: number): number {
  const scaled = Math.round((rate / 0.3) * 100);
  return Math.max(0, Math.min(100, scaled));
}

function sortedTotals(totals: Map<string, number>): KeyTotal[] {
  return [...totals.entries()]
    .map(([key, total]) => ({ key, total }))
    .sort((a, b) => b.total - a.total || a.key.localeCompare(b.key));
}

export function buildReport(txs: ReportTx[], largestN = 5): Report {
  const months = new Map<string, MonthPoint>();
  const byCategory = new Map<string, number>();
  const byMerchant = new Map<string, number>();
  let incomeTotal = 0;
  let expenseTotal = 0;
  const expenses: ReportTx[] = [];

  for (const tx of txs) {
    if (tx.type === "transfer") continue;
    const month = tx.date.slice(0, 7);
    const point = months.get(month) ?? { month, income: 0, expense: 0, net: 0 };
    if (tx.type === "income") {
      incomeTotal += tx.amount;
      point.income += tx.amount;
    } else {
      expenseTotal += tx.amount;
      point.expense += tx.amount;
      expenses.push(tx);
      if (tx.categoryId) byCategory.set(tx.categoryId, (byCategory.get(tx.categoryId) ?? 0) + tx.amount);
      if (tx.merchantId) byMerchant.set(tx.merchantId, (byMerchant.get(tx.merchantId) ?? 0) + tx.amount);
    }
    point.net = point.income - point.expense;
    months.set(month, point);
  }

  const rate = savingsRate(incomeTotal, expenseTotal);
  return {
    byMonth: [...months.keys()].sort().map((m) => months.get(m)!),
    spendingByCategory: sortedTotals(byCategory),
    spendingByMerchant: sortedTotals(byMerchant),
    largestExpenses: [...expenses]
      .sort((a, b) => b.amount - a.amount || a.date.localeCompare(b.date))
      .slice(0, largestN),
    incomeTotal,
    expenseTotal,
    net: incomeTotal - expenseTotal,
    savingsRate: rate,
    healthScore: healthScore(rate),
  };
}
